Players.hs

```diff
@@ -65,3 +66,56 @@ argmax (h:t)
    where
       (bt,vt) = argmax t
       vh = snd h
+
+-- MiniMax Player --
+type Mem = Dict State (Action, Double)
+
+-- mmPlayer plays out the moves using the MiniMax algorithm
+-- Unfortunately because of the huge amount of moves, it takes forever when there are a lot of possible moves remaining
+mmPlayer :: Player
+mmPlayer st = do
+    return (fst (fst (minimax connect4 st emptyDict)))
+
+----   Determining the best move  ---
+----   These functions were taken from CPSC 312 lecture files ----
+minimax:: Game -> State -> Mem -> ((Action, Double), Mem)
+-- minimax game state memory  =>  ((move,value_to_player),new_memory)
+minimax game st mem =
+   case getval st mem of
+      Just act_val  -> (act_val,mem)
+      Nothing ->
+        let (act_val,mem1) =
+              argmax_mem (valueact game st) avail mem
```

Comment:
Variable not in scope:
        argmax_mem
          :: (Action -> Mem -> (Double, Mem)) -> [Action] -> Mem -> (a, b)
    * Perhaps you meant `argmaxMem' (line 106)

Reply:
good catch